provider/cloudflare/cloudflare.go

```diff
@@ -556,20 +697,31 @@ func groupByNameAndType(records []cloudflare.DNSRecord) []*endpoint.Endpoint {
 		groups[groupBy] = append(groups[groupBy], r)
 	}
 
+	// map custom origin to custom hostname, custom origin should match to a dns record
+	customOriginServers := map[string]string{}
+
+	// only one latest custom hostname for a dns record would work
+	for _, c := range chs {
+		customOriginServers[c.CustomOriginServer] = c.Hostname
+	}
+
 	// create single endpoint with all the targets for each name/type
 	for _, records := range groups {
 		targets := make([]string, len(records))
 		for i, record := range records {
 			targets[i] = record.Content
 		}
-		endpoints = append(endpoints,
-			endpoint.NewEndpointWithTTL(
-				records[0].Name,
-				records[0].Type,
-				endpoint.TTL(records[0].TTL),
-				targets...).
-				WithProviderSpecific(source.CloudflareProxiedKey, strconv.FormatBool(*records[0].Proxied)),
-		)
+		ep := endpoint.NewEndpointWithTTL(
```

Comment:
it is possible for `records[0]` to cause an out-of-index exception if the records slice is empty.

we most likely do need 
```go
if len(records) == 0 {
    return endpoints
}
```

Reply:
added
(applicable to the existing code as well)